### DISCUSSION TITLE
Adapter Maintainers: Upgrading to dbt-core 1.4.0

### DISCUSSION THREAD

**Discussion by dataders** [dbt-labs/dbt-core#6624] [Community]
> **TL;DR** Not much heavy lifting for this minor version. But FYI there will be more work for `1.5.0`. Be warned!

This discussion is for communicating to adapter maintainers the scope of work needed to make use of the changes in 1.4.0. If you have questions and concerns, please ask them here for posterity.

For the guide to upgrading an adapter to be compatible with dbt-core=1.3.0, please see #6011.

The latest release cut for 1.4.0, `dbt-core==1.4.0rc1`, was published on January 11, 2023 ([PyPI](https://pypi.org/project/dbt-core/1.4.0rc1/) | [Github](https://github.com/dbt-labs/dbt-core/releases/tag/v1.4.0rc1)). We are targeting releasing the official cut of 1.4.0 in time for January 25, 2023. We ask that adapter maintainers strive to release an official version of corresponding adapter package by February 8, 2023.

## Upgrade dependencies

Be sure to also upgrade the version of `dbt-tests-adapter`

```pip
# for all package users
# dbt-core~=1.4.0rc1 until final release
dbt-core~=1.4.0

# for adapter tests from core
# dbt-tests-adapter~=1.4.0rc1
dbt-tests-adapter~=1.4.0 until final release
```

## Add support for Python 3.11

The long wait is over for [Guido's MSFT-funded foray into speeding up his pet project](https://devblogs.microsoft.com/python/python-311-faster-cpython-team/). However these performance improvements are not immediately realizble in dbt until key package dependencies first start supporting (and testing against) Python 3.11. As of the time of writing (1/16/23), *most* Python database driver libraries/packages support Python 3.11

If possible, please update your adapter to support Python 3.11 and also test against it. Verified adapters must test against this version, and adapters included in cloud must test against all currently supported minor version of the Python 3 (i.e. 3.7 though 3.11).

## Consolidate timestamp functions & macros

Most of the work was done in core (https://github.com/dbt-labs/dbt-core/issues/5521 https://github.com/dbt-labs/dbt-bigquery/pull/323/files). Feel free to dive in to the PRs and related issues, but my TL;DR is that just because macros had the same name across adapters didn't mean the implementation was the same or was guaranteed to return the same datatype. The consensus is now to standardize timestamp-related macros to make use of the convert_timezone() macro and built in Python logic.

There is a new adapter-zone test to implement to ensure things are hunky-dory:
`dbt.tests.adapter.utils.test_timestamps.BaseCurrentTimestamps`

Any adapter-specific override of the following time-related macros should now live in `macros/utils/timestamps.sql` 
    - `current_timestamp`
    - `snapshot_string_as_time`
    - `snapshot_get_time`

Relatedly, if you maintain a shim of dbt-utils that overrides these macros, changes similar to https://github.com/dbt-labs/dbt-utils/pull/694 may be required

## Replace deprecated exception functions

With the exception to some logging upgrades and small tweaks, dbt-core's Exceptions library ([core/dbt/exceptions.py](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/exceptions.py)) has sat largely dormant for nearly three years. Strong exception handling interface is key to a solid API upon which other teams and tools can rely.

https://github.com/dbt-labs/dbt-core/pull/6347/
https://github.com/dbt-labs/dbt-snowflake/issues/366

`RuntimeException` -> `DbtRuntimeError`
`DatabaseException` -> `DbtDatabaseError`
`raise_compiler_error()` -> `CompilationError`

## Incremental Predicates

[Incremental Predicates](https://docs.getdbt.com/docs/build/incremental-models#about-incremental_predicates) is a new feature for 1.4. See https://github.com/dbt-labs/dbt-core/pull/5702 for more information

There are also tests (`TestIncrementalPredicatesDeleteInsert`, `TestPredicatesDeleteInsert` )that can be used to validate the implementation if your adapter supports either of the mentioned incremental strategies.

### Incremental Tests

If your adapter has a custom incremental "unique key" strategy and/or materializaiton `TestIncrementalUniqueKey` is a recommended approach
https://github.com/dbt-labs/dbt-core/pull/4958

### Query Comment Tests

If your data platform supports query comments, then https://github.com/dbt-labs/dbt-core/pull/5971 migrates the related tests to the adapter zone for you to implement (see https://github.com/dbt-labs/dbt-redshift/pull/197).

### TestChangeRelationTypes

TODO explain the value of this test.... @colin-rogers-dbt? https://github.com/dbt-labs/dbt-core/pull/6054 https://github.com/dbt-labs/dbt-redshift/pull/203

## Example adapter diffs

https://github.com/dbt-labs/dbt-redshift/compare/v1.3.0...v1.4.0rc1
https://github.com/dbt-labs/dbt-bigquery/compare/v1.3.0...v1.4.0rc1
https://github.com/dbt-labs/dbt-snowflake/compare/v1.3.0...v1.4.0rc1
https://github.com/dbt-labs/dbt-spark/compare/v1.3.0...v1.4.0rc1
https://github.com/dbt-labs/dbt-bigquery/compare/v1.3.0...v1.4.0rc1

**Comment by morsapaes** [comment]
Thanks for the overview (as usual), @dataders! 

There is a pending question from our side in terms of test coverage: is there a recommendation around what tests **must** be included in each external adapter's test suite? Since `dbt-materialize` tests run in our main CI pipeline and the adapter largely shims `dbt-postgres`, we'd like to keep things slim if things can be kept slim!

**Reply by dataders** [reply]
fair question. I can't speak to y'all's optimization tradeoffs. that's a "you problem" 😺 .

A confounding factor here is that this suite aren't unit tests, but rather integration tests in that they're useless without a data platform on which to run them. In theory you could just run the subset of `adapter/` zone tests that are distinct from dbt-postgres's, but, what assurances do we have that the dbt-postgres code used as-in on Materialize's database will work?

See that all relevant tests have passed for each release is a requirement to be a "verified" core adapter. But that's only for the release commits, so perhaps there's tweaking of the GitHub actions such that the full suite runs only on commits associated with releases? otherwise it is the subset? 🤷🏻 

That said, we're always looking to improve the value and UX of our adapter testing suite, so if there's specific feedback about tests that are too slow, let us know!

**Comment by genzgd** [comment]
Is there discussion of this change somewhere?

```python
quote_policy: SnowflakeQuotePolicy = field(default_factory=lambda: SnowflakeQuotePolicy())
```

Are all policies required to be a `dataclasses.field` now?

**Reply by colin-rogers-dbt** [reply]
This wasn't a change we _wanted_ to make but a requirement for supporting py3.11, see: https://github.com/dbt-labs/dbt-core/pull/6326

**Reply by genzgd** [reply]
Ah, that makes sense for dataclasses, thanks!

**Comment by nicor88** [comment]
When attempting to upgrade the dbt-athena adapter (PR: https://github.com/dbt-athena/dbt-athena/pull/134) I'm having few issues:
- `caplog.text` used in our unit testing  (with pytest) to catch some errors/messages when in debug mode, started to be empty (it was fully working on version 1.3.x) - could be that dbt is now flushing the caplog therefore is empty. Could be that `logger.propagate = False` on the logs are not push down.
- ~some of our functions tests started to fail~
  <pre>
  FAILED tests/functional/adapter/utils/test_utils.py::TestArrayAppend::test_expected_actual - TypeError: '_MISSING_TYPE' object is not callable
  FAILED tests/functional/adapter/utils/test_utils.py::TestArrayConstruct::test_expected_actual - TypeError: '_MISSING_TYPE' object is not callable
  FAILED tests/functional/adapter/utils/test_utils.py::TestArrayConcat::test_expected_actual - TypeError: '_MISSING_TYPE' object is not callable
  FAILED tests/functional/adapter/utils/test_utils.py::TestExcept::test_build_assert_equal - TypeError: '_MISSING_TYPE' object is not callable
  FAILED tests/functional/adapter/utils/test_utils.py::TestIntersect::test_build_assert_equal - TypeError: '_MISSING_TYPE' object is not callable
  </pre>
  ~Seems due to~
  <pre>
  @classmethod
    def get_default_include_policy(cls) -> Policy:
  >       return cls._get_field_named("include_policy").default_factory()
  E       TypeError: '_MISSING_TYPE' object is not callable
  </pre>
   ~Any clue on what to fix here?~  
   Fixed adding : 
   <pre>
   @dataclass(frozen=True, eq=False, repr=False)
   class AthenaRelation(BaseRelation):
      quote_character: str = ""
      include_policy: Policy = field(default_factory=lambda: AthenaIncludePolicy())
   </pre>

Overall the core functionalities are working, but the above is preventing us to move further with merging and releasing to the new version.

**Reply by genzgd** [reply]
For the missing type do something like this in your Relation subclass:

```python
@dataclass(frozen=True, eq=False, repr=False)
class ClickHouseRelation(BaseRelation):
    quote_policy: Policy = field(default_factory=lambda: ClickHouseQuotePolicy())
    include_policy: Policy = field(default_factory=lambda: ClickHouseIncludePolicy())
```

**Reply by genzgd** [reply]
I had the same problem prompted my comment above:  https://github.com/dbt-labs/dbt-core/discussions/6624#discussioncomment-4853492

**Reply by nicor88** [reply]
~I tried with one of the failing tests and seems that worked.~ All our functional testings are now working again.

**Reply by dataders** [reply]
looks like this finding is worth adding the the main upgrade guide -- either of y'all mind writing up a snippet and I'll add it in?

**Reply by nicor88** [reply]
I think that you can pick what @genzgd posted
```python
from dataclasses import dataclass, field

@dataclass(frozen=True, eq=False, repr=False)
class ClickHouseRelation(BaseRelation):
    quote_policy: Policy = field(default_factory=lambda: ClickHouseQuotePolicy())
```

**Comment by ravjotbrar** [comment]
Is the new [tests/adapter/dbt/tests/adapter/utils/test_timestamps.py](https://github.com/dbt-labs/dbt-core/pull/5838/files#diff-649f9b4416cee71ccd51b3b9cfce09b374d393e2a125d7423c9a428928d7bbaf) test only applicable to postgres and snowflake adapters? I see this in the changes for the relevant macros in timestamp.sql:
```
/* {#
    DEPRECATED: DO NOT USE IN NEW PROJECTS

    This is ONLY to handle the fact that Snowflake + Postgres had functionally
    different implementations of {{ dbt.current_timestamp }} + {{ dbt_utils.current_timestamp }}

    If you had a project or package that called {{ dbt_utils.current_timestamp() }}, you should
    continue to use this macro to guarantee identical behavior on those two databases.
#} */
```

**Comment by dataders** [comment]
here's current progress


- [x] https://github.com/ClickHouse/dbt-clickhouse/issues/135
- [x] https://github.com/databricks/dbt-databricks/issues/267
- [x] https://github.com/jwills/dbt-duckdb/issues/88
- [x] https://github.com/tglunde/dbt-exasol/issues/57
- [x] https://github.com/firebolt-db/dbt-firebolt/issues/93
- [x] https://github.com/aws-samples/dbt-glue/issues/143
- [x] https://github.com/markporoshin/dbt-greenplum/issues/14
- [x] https://github.com/cloudera/dbt-hive/issues/94
- [x] https://github.com/aurany/dbt-ibmdb2/issues/30
- [x] https://github.com/cloudera/dbt-impala/issues/142
- [x] https://github.com/MaterializeInc/materialize/issues/17500
- [x] https://github.com/dbeatty10/dbt-mysql/issues/122
- [x] https://github.com/oracle/dbt-oracle/issues/70
- [x] https://github.com/memsql/dbt-singlestore/issues/18
- [x] https://github.com/dbt-msft/dbt-sqlserver/issues/333
- [x] https://github.com/dbt-athena/dbt-athena/issues/136
- [x] https://github.com/microsoft/dbt-synapse/issues/147
- [x] https://github.com/starburstdata/dbt-trino/issues/223
- [x] https://github.com/vertica/dbt-vertica/issues/67
- [x] https://github.com/dremio/dbt-dremio/issues/117
- [x] https://github.com/fal-ai/fal/issues/740
- [x] https://github.com/databendcloud/dbt-databend/issues/1
- [x] https://github.com/Teradata/dbt-teradata/issues/63
- [ ] https://github.com/iomete/dbt-iomete/issues/5
- [ ] https://github.com/mindsdb/dbt-mindsdb/issues/18
- [ ] https://github.com/chop-analytics/dbt-netezza/issues/18
- [ ] https://github.com/rockset/dbt-rockset/issues/28
- [ ] https://github.com/splitgraph/sgr/issues/773
- [ ] https://github.com/codeforkjeff/dbt-sqlite/issues/40
- [ ] https://github.com/pingcap/dbt-tidb/issues/49
- [ ] https://github.com/inferlabs/dbt-infer/issues/5
- [ ] https://github.com/StarRocks/dbt-starrocks/issues/15
